services/galley/test/integration/API/MLS.hs

```diff
@@ -670,6 +705,134 @@ testCommitNotReferencingAllProposals = withSystemTempDirectory "mls" $ \tmp -> d
   err <- testFailedCommit (MessagingSetup {creator = alice, users = [bob, dee], ..}) 409
   liftIO $ Wai.label err @?= "mls-commit-missing-references"
 
+testAdminRemovesUserFromConv :: TestM ()
+testAdminRemovesUserFromConv = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" (pClients bob)
+
+  do
+    convs <-
+      responseJsonError =<< getConvs (qUnqualified (pUserId bob)) Nothing Nothing
+        <!! const 200 === statusCode
+    liftIO $
+      assertBool
+        "bob is in conversation before he gets removed"
+        (conversation `elem` map cnvQualifiedId (convList convs))
+
+  testSuccessfulRemoveMemberFromConvCommit creator [bob] conversation removalCommit [bob]
+
+  do
+    convs <-
+      responseJsonError =<< getConvs (qUnqualified (pUserId bob)) Nothing Nothing
+        <!! const 200 === statusCode
+    liftIO $
+      assertBool
+        "bob is not longer part of conversation after the commit"
+        (conversation `notElem` map cnvQualifiedId (convList convs))
+
+testRemoveClientsIncomplete :: TestM ()
+testRemoveClientsIncomplete = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  -- remove only first client of bob
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" [(NE.head (pClients bob))]
+
+  err <-
+    responseJsonError
+      =<< postMessage (qUnqualified (pUserId creator)) removalCommit
+        <!! statusCode === const 409
+
+  liftIO $ Wai.label err @?= "mls-client-mismatch"
+
+testUserRemovesThemselvesFromConv :: TestM ()
+testUserRemovesThemselvesFromConv = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
```

Comment:
Isn't this implemented by `aliceInvitesBob`?

Reply:
Right, I've updated it